sklearn/utils/extmath.py

```diff
@@ -689,28 +689,41 @@ def _incremental_mean_and_var(X, last_mean=.0, last_variance=None,
     # new = the current increment
     # updated = the aggregated stats
     last_sum = last_mean * last_sample_count
-    new_sum = X.sum(axis=0)
+    sum_func = np.nansum if ignore_nan else np.sum
+    new_sum = sum_func(X, axis=0)
+    new_sum[np.isnan(new_sum)] = 0
 
-    new_sample_count = X.shape[0]
+    new_sample_count = np.sum(~np.isnan(X), axis=0)
     updated_sample_count = last_sample_count + new_sample_count
 
+    warnings.filterwarnings('ignore')  # as division by 0 might happen
     updated_mean = (last_sum + new_sum) / updated_sample_count
+    updated_mean[np.isinf(updated_mean)] = 0
+    updated_mean[np.isnan(updated_mean)] = 0
 
     if last_variance is None:
         updated_variance = None
     else:
-        new_unnormalized_variance = X.var(axis=0) * new_sample_count
-        if last_sample_count == 0:  # Avoid division by 0
-            updated_unnormalized_variance = new_unnormalized_variance
-        else:
-            last_over_new_count = last_sample_count / new_sample_count
-            last_unnormalized_variance = last_variance * last_sample_count
-            updated_unnormalized_variance = (
-                last_unnormalized_variance +
-                new_unnormalized_variance +
-                last_over_new_count / updated_sample_count *
-                (last_sum / last_over_new_count - new_sum) ** 2)
+        var_func = np.nanvar if ignore_nan else np.var
+        new_unnormalized_variance = var_func(X, axis=0)
+        new_unnormalized_variance[np.isnan(new_unnormalized_variance)] = 0
+        new_unnormalized_variance *= new_sample_count
+        last_over_new_count = last_sample_count / new_sample_count
+        last_unnormalized_variance = last_variance * last_sample_count
+        updated_unnormalized_variance = (
+            last_over_new_count / updated_sample_count *
+            (last_sum / last_over_new_count - new_sum) ** 2)
+        # updated_unnormalized_variance can be both NaN or Inf
+        updated_unnormalized_variance[np.isnan(
```

Comment:
can we just use `updated_unnormalized_variance[np.logical_not(new_sample_count)] = 0` or something similar?

Reply:
@jnothman yeah sure, but not exactly `new_sample_count` but `updated_unnormalized_variance[np.logical_not(updated_sample_count)] = 0` because consider the case:
- `last_sample_count = [1, 3, 5, 7, 9]`
- `new_sample_count = [3, 1, 0, 5, 9]'
- so `updated_unnormalized_variance[2] != 0` just because `new_sample_count[2] == 0`

Reply:
Fine. Seems simpler than isnan and isinf to me